src/System.Data.Jet/ComObject.cs

```diff
@@ -0,0 +1,134 @@
+        }
+
+        public ComObject(Guid clsid)
+            : this(Activator.CreateInstance(Type.GetTypeFromCLSID(clsid, true)))
+        {
+        }
+
+        public object Detach()
+        {
+            var instance = _instance;
+            _instance = null;
+            return instance;
+        }
+
+        public override bool TryGetMember(GetMemberBinder binder, out object result)
+        {
+            result = WrapIfRequired(
+                _instance.GetType()
+                    .InvokeMember(
+                        binder.Name,
+                        BindingFlags.GetProperty,
+                        Type.DefaultBinder,
+                        _instance,
+                        new object[0]
+                    ));
+            return true;
+        }
+
+        public override bool TrySetMember(SetMemberBinder binder, object value)
+        {
+            _instance.GetType()
+                .InvokeMember(
+                    binder.Name,
+                    BindingFlags.SetProperty,
+                    Type.DefaultBinder,
+                    _instance,
+                    new[]
+                    {
+                        value is ComObject comObject
+                            ? comObject._instance
+                            : value
+                    }
+                );
+            return true;
+        }
+
+        public override bool TryInvokeMember(InvokeMemberBinder binder, object[] args, out object result)
+        {
+            result = WrapIfRequired(
+                _instance.GetType()
+                    .InvokeMember(
+                        binder.Name,
+                        BindingFlags.InvokeMethod,
+                        Type.DefaultBinder,
+                        _instance,
+                        args
+                    ));
+            return true;
+        }
+
+        public override bool TryGetIndex(GetIndexBinder binder, object[] indexes, out object result)
+        {
+            // This should work for all specific interfaces derived from `_Collection` (like `_Tables`) in ADOX.
+            result = WrapIfRequired(
+                _instance.GetType()
+                    .InvokeMember(
+                        "Item",
+                        BindingFlags.GetProperty,
+                        Type.DefaultBinder,
+                        _instance,
+                        indexes
+                    ));
+            return true;
+        }
+
+        // See https://github.com/dotnet/runtime/issues/12587#issuecomment-578431424
+        private static object WrapIfRequired(object obj)
+        {
+            if (obj != null && !obj.GetType()
+                .IsPrimitive)
+            {
+                return new ComObject(obj);
+            }
+
+            return obj;
+        }
+        
+        private void ReleaseUnmanagedResources()
+        {
+            if (_instance != null)
+            {
+                Marshal.FinalReleaseComObject(_instance);
```

Comment:
Well, the important aspect here is, that this class is intended as a wrapper for the lifetime of the underlying COM object.

So as long as we never directly expose the underlying object, but only the `ComObject` class, this is actually a safer way to implement this, because most people nowadays are not familiar with the COM concepts and could make mistakes that could lead to the object staying alive in the process. To mitigate this, `Marshal.FinalReleaseComObject()` will definitely release the underlying object, when the `ComObject` wrapper gets disposed.

And in case we ever want to expose the underlying object to the library caller for some reason, I added the (currently unused) `Detach()` method, that will hand the underlying object over to whoever wants to continue use it.

Reply:
I'm not sure I can explain this well, but multiple RCWs can essentially refer to the same unmanaged memory, so I believe a call to `Marshal.FinalReleaseComObject()` can also corrupt an object that you did not create (e.g. an object that a user created outside of the library).  It is possible that this specific situation avoids that (I don't think so, but I'm not 100% sure).  However, I think my main point is that you don't have to share the internal ComObject reference with consumer code to end up with errors.

Probably the best thing that I could do would be to write a simple test and see if I get an AV or other exception as a result of `Marshal.FinalReleaseComObject()` being called internally instead of `Marshal.ReleaseComObject()`.

Reply:
That being said, I think it's pretty unlikely that people would be using COM interop in any way.  They are most likely using EFCore.Jet to avoid things like that.

Reply:
The only point in our code we use this class is to handle the dynamic ADOX scenarios. We never expose the class to the outside.
The RCW does not actually makes sure, that the underlying COM object is actually destroyed. The RCW only "keeps" one reference count to the underlying COM object at all times. Any other reference count increments are handled internally by the RCW, but are not delegated to the underlying COM object.

So the difference between `ReleaseComObject` and `FinalReleaseComObject()` is just, that `ReleaseComObject` does decrement the RCWs *internal* reference counter by one, and if it reaches zero, also calls `IUnknown->Release()` on the underlying COM object, while `FinalReleaseComObject` does decrement the RCWs *internal* reference counter always until it is zero, and then calls `IUnknown->Release()` on the underlying COM object.

That does not mean, that the underlying COM object will be destroyed. As long as it's own reference counter is still greater than zero, it will still exist. There is no way for the COM object to be prematurely released in the ways I have implemented it.

The only thing that we could screw up is, that we expose the object to the outside world, but forget to detach it. But it would not make any difference then, if we call `FinalReleaseComObject` or `ReleaseComObject` on the RCW, because its internal ref count would be one, both either method would decrement it to zero, and the RCW would dispose itself (the underlying object only if also the ref count of the underlying object would reach zero, which would likely to be the case here).

But this is as safely as this can be implemented. I developed COM components between 1995 and 2006, so I am actually a COM expert. If you have any further questions about COM or COM-Interop, feel free to discuss this further.

Reply:
@lauxjpn Good to know.  I thought I was going insane for a second.  I was about to swear that I read something in docs & witnessed something in System.Data.OleDb that led to me to believe what I was believing.  I am sorry that I probably wasn't using the correct terminology and wasn't very good at explaining it.

Reply:
No, it's my fault. It has been a long day for me and I tend to fall back into bad habits when I am very tired.